ddtrace/tracer/payload.go

```diff
@@ -160,3 +192,131 @@ func (p *payload) Read(b []byte) (n int, err error) {
 	}
 	return p.reader.Read(b)
 }
+
+// Write implements io.Writer. It writes data directly to the buffer.
+func (p *unsafePayload) Write(data []byte) (n int, err error) {
+	return p.buf.Write(data)
+}
+
+// grow grows the buffer to ensure it can accommodate n more bytes.
+func (p *unsafePayload) grow(n int) {
+	p.buf.Grow(n)
+}
+
+// recordItem records that an item was added and updates the header.
+func (p *unsafePayload) recordItem() {
+	atomic.AddUint32(&p.count, 1)
+	p.updateHeader()
+}
+
+// stats returns the current stats of the payload.
+func (p *unsafePayload) stats() payloadStats {
+	return payloadStats{
+		size:      p.size(),
+		itemCount: int(atomic.LoadUint32(&p.count)),
+	}
+}
```

Comment:
@kakkoyun Aren't we going to allocate these small structs over and over? I saw that we are keeping the `size/itemCount` methods.

Reply:
Unfortunately, we need to do that. To actually see the overhead we need to measure. I can do that. And maybe do some tricks. Since all these allocations are local I expect the overhead to be minimal.

Reply:
Let's try a few tricks, just to make sure we are merging the best version possible. The overhead might be small but I think this is escaping to the heap, so I worry about GC here.